pkg/controllers/dependencies.go

```diff
@@ -184,9 +208,134 @@ func (l LocalHelmReconciler) Reconcile(ctx context.Context, request helm.ChartRe
 		l.logger.Info("update helm values", "old", vals, "new", request.Values)
 	}
 
+	if strings.Contains(request.ReleaseName, Etcd) {
+		oldSize := vals["persistence"].(map[string]interface{})["size"].(string)
+		newSize := request.Values["persistence"].(map[string]interface{})["size"].(string)
+
+		if parseSize(newSize) != parseSize(oldSize) {
+			l.logger.Info("reconcile PVC", "old size:", oldSize, "new size:", newSize, "release", request.ReleaseName)
+			if err := l.reconcilePVCs(ctx, request.Namespace, request.ReleaseName, oldSize, newSize); err != nil {
+				return err
+			}
+		}
+	}
+
 	return helm.Update(cfg, request)
 }
 
+func parseSize(size string) int64 {
+	size = strings.TrimSpace(size)
+	if len(size) == 0 {
+		return 0
+	}
+
+	unit := size[len(size)-2:]
+	value, err := strconv.ParseInt(size[:len(size)-2], 10, 64)
+	if err != nil {
+		return 0
+	}
+
+	switch strings.ToLower(unit) {
+	case "ti":
+		return value * 1024 * 1024 * 1024 * 1024
+	case "gi":
+		return value * 1024 * 1024 * 1024
+	case "mi":
+		return value * 1024 * 1024
+	case "ki":
+		return value * 1024
+	default:
+		return value
+	}
+}
+
+func (l *LocalHelmReconciler) reconcilePVCs(ctx context.Context, namespace, releaseName, oldSize, newSize string) error {
+	l.logger.Info("Reconciling PVCs", "namespace", namespace, "release", releaseName, "oldSize", oldSize, "newSize", newSize)
+
+	// 1. Get the old StatefulSet
+	stsName := releaseName
+	oldSts, err := l.clientset.AppsV1().StatefulSets(namespace).Get(ctx, stsName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get old StatefulSet: %v", err)
+	}
+
+	newQuantity, err := resource.ParseQuantity(newSize)
+	if err != nil {
+		return fmt.Errorf("failed to parse new size: %v", err)
+	}
+
+	// 2. Update all related PVCs
+	for i := 0; i < int(*oldSts.Spec.Replicas); i++ {
+		pvcName := fmt.Sprintf("data-%s-%d", releaseName, i)
+		pvc, err := l.clientset.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, pvcName, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get PVC %s: %v", pvcName, err)
+		}
+
+		pvc.Spec.Resources.Requests[corev1.ResourceStorage] = newQuantity
+		_, err = l.clientset.CoreV1().PersistentVolumeClaims(namespace).Update(ctx, pvc, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to update PVC %s: %v", pvcName, err)
+		}
+		l.logger.Info("Updated PVC size", "pvc", pvcName, "newSize", newSize)
+	}
+
+	// 3. Recreate the StatefulSet
+	if err := l.recreateStatefulSet(ctx, namespace, stsName, oldSts, newQuantity); err != nil {
+		return err
+	}
+
+	l.logger.Info("Successfully resized PVCs and recreated StatefulSet", "namespace", namespace, "release", releaseName)
+	return nil
+}
+
+func (l *LocalHelmReconciler) recreateStatefulSet(ctx context.Context, namespace, name string, oldSts *appsv1.StatefulSet, newQuantity resource.Quantity) error {
```

Comment:
There's a race condition, that when old STS gets deleted, if operator also gets restarted, or failed in wait / create ne w procedure, it won't be able to recover the STS. So we need to add a new field to store the STS status. Or another choice is to not change STS at all to make it simpler

Reply:
Personally, I perfer the second one

Reply:
STS not allow to modify the storage size, so recreate STS is common means to achieve. If not change the STS, only change pvc size. It will be more inelegant. And data is inconsistent

Reply:
@Sallery-X Agree. Then we have to solve the problem of idempotent.  Here's my sugestion:

In the recreate func, we should use `Get` API to check the STS's existence:
- if already deleted, we only need to create.
- if exists, but storage size already updated, we do nothing and continue.
- if exists, and storage size not updated, we do the delete & create procedure. 

Does it make sense to you?

Reply:
Agree. So we have save old STS before deleting.  Save old STS to configmap? secret? or other place?

Reply:
I suggest using ControllerRevision. A configmap is also fine.  If you're to use ControllerRevision, you may refer to the interface here https://github.com/zilliztech/milvus-operator/blob/750dc35b0ef143cc99faf30e9d245d032e74ea56/pkg/controllers/deploy_ctrl_util.go#L59-L62, which is implemented by https://github.com/zilliztech/milvus-operator/blob/750dc35b0ef143cc99faf30e9d245d032e74ea56/pkg/controllers/rollout_util.go#L23-L27

Reply:
ok

Reply:
please review, I have test on my own env @haorenfsa